### DISCUSSION TITLE
Debug tree enhancements

### DISCUSSION THREAD

**Discussion by bitbrain** [bitbrain/beehave#141] [Ideas]
![Capture](https://user-images.githubusercontent.com/822035/219509011-66e1bd95-04f8-4db0-b0f8-9da8e64e264f.PNG)

The tree implemented in #1 works great, however, especially for larger trees it becomes less useful. There are a few challenges here:

- with larger trees, it is a bit difficult to understand what is actually happening (as effectively 50+ actions get activated constantly so it defeats the purpose)
- with larger trees, the rendering of the tree seems to prefer spreading the tree horizontally rather than utilizing the entire space available

In order to solve this, I am proposing a few changes:

1. update the rendering logic so trees are rendered in a more compacted manner by moving some branches further down rather letting them on the same "level"
2. have an option to collapse/expand sections of a tree
3. have a way to "slow down" the ticks somehow - this might be difficult and maybe can be solved separately by having an ingame speed scale applied instead rather than solving it in this addon?
4. have a toggle to show more information on nodes (like variable values etc.)
5. have a way to also visualise blackboards?

I am curious to hear your thoughts!

**Comment by GeroVeni** [comment]
Since the nodes in general are mostly horizontal, I wonder if a left-to-right direction for the tree would fit it better the the current top-to-bottom. This could still benefit from further branch placement optimizations but I think would make the graph more compact

**Reply by DarkMessiah** [reply]
<details>
<summary>Horizontal layout</summary>

https://user-images.githubusercontent.com/1756388/219956908-78d0f2ab-e3c0-4b21-82d0-295e9915b9f8.mp4
</details>

**Reply by Werninator** [reply]
This is probably the way to go, +1!

**Comment by Werninator** [comment]
I'm putting a +1 on proposal 2: allowing to fold a sub-tree by clicking on a node and maybe giving them a theoretical name that is not reflected in data might be a cool way to simplify things. And maybe even help to encapsulate certain sub-trees with visual feedback before actually moving things around.

Another approach would maybe be to have some kind of filter to only show nodes that have returned a state already and/or hide nodes that haven't been ticked for a while. This might be confusing if the speed is too high, though as things would pop in and pop out in large trees.

**Reply by DarkMessiah** [reply]
What do you think about this?

`Click` - Minimize/Maximize node
`Ctrl + Click` - Minimize/Maximize node and children

<details>

https://user-images.githubusercontent.com/1756388/219957226-2a5843f6-f99d-4842-8381-fb3a2e5af9fe.mp4

</details>

**Reply by Werninator** [reply]
Oh this is already helpful, very cool!

I was thinking something like this, plus maybe being able to add a title to a minimized Node:
<details>
<summary>img</summary>

![image](https://user-images.githubusercontent.com/6310506/219957834-8ca5b84f-cd3d-459d-b0a8-62447d954b6c.png)
</details>

But maybe you have an even better approach, as you can still see the type type of nodes

**Reply by DarkMessiah** [reply]
Hm, I think we can impement both options :D
* Minimize/Maximize for all nodes
* Collapse/Uncollapse for `Composite` or `Decorator` nodes
